Upgrading to Entities 1.0.16

jkeon opened the pull request:
Upgrading the code base to Entities 1.0.16

### What is the current behaviour?

Currently on 0.51

### What is the new behaviour?

- Removed all `[BurstCompatible]` and `[NotBurstCompatible]` attributes.
    - These were actually unit test directives and didn't interface with the Burst compiler.
-  No longer allowed to use `struct` for most generic constraints. Must use `unmanaged`
- Types use `TypeIndex` instead of `int`s
- `WorldInternal` no longer has `OnWorldCreated` or `OnWorldDestroyed` events. 
    - Applications can wrap their world creation/destruction to a central location and dispatch their own events or subclass `World` and handle any specific logic in the Constructor/Dispose method.
- `Translation`, `Rotation`, and `Scale` components no longer exist, instead a `LocalTransform` component is used.
- Updated Assembly Definitions to remove unneeded references.
- All `World.GetOrCreateSystem` and related API's are now `World.GetOrCreateSystemManaged` and related APIs.
- Removed `DynamicBufferSharedWriteHandle` and the associated bloat that was necessary with `WorldCache`.
    - This wasn't being used and in practice was better handled by `DBFEWriter`
- `BufferFromEntity` is now `BufferLookup`
- `ComponentDataFromEntity` is now `ComponentLookup`
- Most APIs from Unity that used to return `NativeArray` now return `NativeList` to account for disabled entities. 
    - Related code is now updated to reflect this.
-  All `IJobEntityBatch` are now `IJobChunk`
-  Reworked `ParallelAccessUtil` since it's now much more simplified. 
- Changed `ITaskForDefer`, `ITaskCancelJobForDefer` and `ITaskUpdateJobForDefer` to no longer be generic jobs.
    - Entities 1.0 uses `ISystem`s for Burstable systems and generic job types are no longer allowed. 
    - In these jobs, there must be a reader declared in the job and fulfilled by the job data.
    - The `Execute` function will give an index which can then be used to get the right element from the reader.
 - Updated README

### What issues does this resolve?
- None
- While #86 identifies a lot of the work that was done in this PR, the TODO's throughout the code refer to additional work that this PR does not address.

### What PRs does this depend on?
 - None

### Does this introduce a breaking change?
 - [x] Yes - Numerous API changes. Projects using this library will also need to migrate application code to Entities 1.0.16
 - [ ] No

jkeon committed:
Initial work to make it compile

jkeon committed:
Merge remote-tracking branch 'origin/main' into upgrade-path

jkeon committed:
Upgrading to 1.0

jkeon committed:
API upgrades

jkeon committed:
Removing warnings

jkeon committed:
WIP

jkeon committed:
Fixing warnings

jkeon committed:
Updating for the new Parallel Access

jkeon committed:
Updating TaskJobForDefer to attempt to get Burst to work

jkeon committed:
Updated ITaskJobForDefer to be de-genericized

jkeon committed:
De-genericizing Update and Cancellation Update jobs

jkeon committed:
Working Burstable job types

jkeon committed:
Ready for PR

jkeon committed:
Updated README

mbaker3 commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskCancelJobForDeferExtension.cs:
```diff
@@ -93,61 +109,36 @@ private static void ValidateReflectionData(IntPtr reflectionData)
+            [BurstDiscard]
+            internal static void Initialize()
             {
-                JobData = jobData;
-                CancellationUpdater = scheduleInfo.CancellationUpdater;
-                NativeThreadIndex = UNSET_NATIVE_THREAD_INDEX;
+                if (JOB_REFLECTION_DATA.Data == IntPtr.Zero)
+                {
+                    JOB_REFLECTION_DATA.Data = JobsUtility.CreateJobReflectionData(
+                        typeof(WrapperJobProducer<TJob>),
+                        typeof(TJob),
+                        (ExecuteJobFunction)Execute);
+                }
```
How do calls to this work if it's being burst discarded?
It seems to generate state that is assumed to exist in other parts of the code.

Are you sure this works when burst isn't a JIT operation and AOT generated during a non-editor build?

jkeon replied:
It's not being called from Burst. It happens as part of the `EarlyJobInit` code gen post processor that Unity runs. On startup it will get initialized from the main thread. 

All the existing Unity jobs like `IJobParallelForDefer` etc have `BurstDiscard` on their `Initialize` functions, so we're just "when in Rome..."ing it in ours.

mbaker3 commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/Wrapper/CDFEAccessWrapper.cs:
```diff
@@ -5,15 +5,15 @@
 namespace Anvil.Unity.DOTS.Entities.TaskDriver
 {
     internal class CDFEAccessWrapper<T> : AbstractAccessWrapper
```
Should we rename to `CDLAccessWrapper`? `ComponentLookupAccessWrapper`? `CDAccessWrapper`?

Based on the existing `DynamicBufferAccessWrapper` maybe `ComponentDataAccessWrapper` is the most appropriate.

mbaker3 replied:
Yea, naming is tough. Regardless of what we decide let's make the change in a new PR so we don't clutter this one further.

I think I prefer Option 2 the most or leaving it as is. "Component Data From Entity" is still a valid description of the behaviour. 

To me, I'm not worried about implying sequential access. The usage of the reader/writer is the biggest hint that it's random access. Getting data off of an Entity key is clearly a lookup operation.

I vote either `ComponentDataReader` or `CDFEReader` with a preference to choose the latter and leave it as is. It's short and descriptive.

This isn't a hard position though and I'm open to other opinions.

jkeon replied:
I've created the issue here: https://github.com/decline-cookies/anvil-unity-dots/issues/303

But I also agree, I'm cool with leaving as CDFE.

mbaker3 commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/JobDataInteraction/DataStreamActiveReader.cs:
```diff
@@ -25,9 +26,10 @@ internal DataStreamActiveReader(NativeArray<EntityKeyedTaskWrapper<TInstance>> a
         {
-            get => m_Active[index].Payload;
+            return ref m_Active.ElementAtReadOnly(index).Payload;
         }
```
Should we be dropping support for the indexer? I know most of the time getting the `ref` is more efficient but there are cases where it's not.
Also, if you want to call a bunch of methods non readonly methods on a mutable type then you'll want to convert this out of a `ref` anyway to prevent defensive copies at each call.

plus, we expose the indexer getter for the writers/updaters. Seems like we should keep it here for the sake of consistency.

mbaker3 commented on Scripts/Runtime/Entities/Transform/TransformUtil.cs:
```diff
@@ -35,19 +35,9 @@ public static void AddMissingStandardComponents(Entity entity, EntityManager ent
             Debug.Assert(entityManager.Exists(entity));
 
-            if (!entityManager.HasComponent<Translation>(entity))
+            if (!entityManager.HasComponent<LocalTransform>(entity))
```
I'm not sure this util method is even required anymore now that it's just one type.
If you just want to make a TODO issue for this that's fine.

jkeon replied:
Added TODO for #302

mbaker3 commented on Scripts/Runtime/Entities/Util/ComponentSystemBaseExtension.cs:
```diff
@@ -143,20 +150,23 @@ public static class ComponentSystemBaseExtension
                 }
+
+                NativeList<SystemHandle> worldGroupSystems = worldGroup.GetAllSystems(Allocator.Temp);
+                if (worldGroupSystems.AsArray().All(systemHandle => systemHandle != system.SystemHandle))
```
I don't think we need to introduce `LINQ` here. You could cache the `system.SystemHandle` at the start of the method and then just use `worldGroupSystems.Contains(systemHandle)`.
(Assuming `NativeList` has a contains method.)

Would avoid allocating the conversion to NativeArray too.

jkeon replied:
Good point, done.

mbaker3 reviewed (commented):
I'm just going to go ahead and mention that #296 was created as part of this PR to make digging up the context easier later.

Thanks for doing this. Looks like a lot of work and some tricky gotchas. This is looking good. Think we just need to make some style/clarity decisions.

---

> Removed all [BurstCompatible] and [NotBurstCompatible] attributes. 

To communicate intent should we replace with our own attribute? Or just use comments?


> While Upgrade to Entities 1.0 #86 identifies a lot of the work that was done in this PR, the TODO's throughout the code refer to additional work that this PR does not address.

Can you check off what is complete in that issue? Or if it makes more sense move the other pending work to new issues.

jkeon commented on Scripts/Runtime/Entities/Lifecycle/EntityLifecycleStatus.cs:
```diff
@@ -243,9 +244,9 @@ private struct UpdateArrivedJob : IJobEntityBatch
             }
 
-            public void Execute(ArchetypeChunk batchInChunk, int batchIndex)
+            public void Execute(in ArchetypeChunk chunk, int unfilteredChunkIndex, bool useEnabledMask, in v128 chunkEnabledMask)
```
No because this is concerned with knowing when Entities exist in the world. Their status of being enabled or not is something that would be checked further downstream by something that cares about it. 

If we checked the flag, we would always miss an entity that entered the world disabled.

mbaker3 replied:
Shouldn't we configure that in the query that this job operates on instead though?

Fair enough if skipping the check is an optimization. We should just have a comment to explain why we're skipping it and mentioning that our query is also requesting all disabled entities.

There's also a risk here that today, or in the future, there is a chunk level optimization where Unity will skip a chunk if all entities in that chunk are disabled and the query isn't requesting to find disabled entities as well.

jkeon replied:
Good point, in reading deep into the docs, if an entire chunk is disabled, it will skip the chunk. 

The query for this will now ensure it includes disabled entities and disabled components.

mbaker3 replied:
Can you make sure to add the comment about why you're ignoring the mask?

jkeon replied:
Yep, done!

jkeon commented on Scripts/Runtime/Entities/Lifecycle/Migration/EntityWorldMigrationSystem.cs:
```diff
@@ -105,6 +105,8 @@ private JobHandle NotifyObserversOfMigrateTo(World destinationWorld, ref NativeA
         private static NativeList<TypeManager.EntityOffsetInfo> s_EntityOffsetList = new NativeList<TypeManager.EntityOffsetInfo>(32, Allocator.Persistent);
         private static NativeList<TypeManager.EntityOffsetInfo> s_BlobAssetRefOffsetList = new NativeList<TypeManager.EntityOffsetInfo>(32, Allocator.Persistent);
         private static NativeList<TypeManager.EntityOffsetInfo> s_WeakAssetRefOffsetList = new NativeList<TypeManager.EntityOffsetInfo>(32, Allocator.Persistent);
+        private static NativeList<TypeManager.EntityOffsetInfo> s_UnityObjectRefOffsetList = new NativeList<TypeManager.EntityOffsetInfo>(32, Allocator.Persistent);
```
Yeah, the next version of Entities (1.1 preview, current version is 1.2 preview) adds this. 

I had originally ported to 1.2 but then downgraded to figure out why our custom jobs weren't working. It turned out to be unrelated so we could go up to 1.2 whenever we need to. 

I've deleted this though so it's not confusing.

jkeon committed:
Removing code that was for a 1.2 entities upgrade but we decided to stay at 1.0.16 for now

jkeon commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobData/CancelJobData.cs:
```diff
@@ -20,13 +20,16 @@ internal CancelJobData(CancelJobConfig<TInstance> jobConfig) : base(jobConfig)
+        /// Fulfills an instance of the provided type for the job.
+        /// </summary>
+        /// <param name="cancellationUpdater">The <see cref="DataStreamCancellationUpdater{TInstance}"/></param>
+        public void Fulfill(out DataStreamCancellationUpdater<TInstance> cancellationUpdater)
```
The request side for this is implicit by the type of job.

`private struct CancelMoveJob : ITaskCancelJobForDefer`

Previously, we would feed your `Execute` function for you. 

`public void Execute(ActorMove pendingCancelInstance)`

But now because we can't use generics in the job interface type, we have to give you the index.

```csharp
public void Execute(int index)
{
                ActorMove pendingCancelInstance = m_CancellationUpdater[index];
}
```

Previously the `m_CancellationUpdater` was hidden from you. 

Now you will do:

```csharp
private DataStreamCancellationUpdater<ActorMove> m_CancellationUpdater;

            private CancelMoveJob(CancelJobData<ActorMove> jobData) : this()
            {
                jobData.Fulfill(out m_CancellationUpdater);
            }
```

Your concern of `RequestCancelFor(target)` is still the exact same because that is about requesting the start of a cancel for a certain task driver. 

This fulfill is for the job that processes elements that need custom cancel logic.

mbaker3 replied:
ah ok cool. I thought this was part of the cancel request stream stuff.

jkeon commented on Scripts/Runtime/Entities/TaskDriver/TaskSet/Job/JobType/TaskCancelJobForDeferExtension.cs:
```diff
@@ -83,6 +84,21 @@ public static class TaskCancelJobForDeferExtension
         //*************************************************************************************************************
 
+        [UsedImplicitly]
+        public static void EarlyJobInit<TJob>()
```
Unity does. 

This is just how jobs are declared and used. Within Unity's code gen, it calls the following:

`methodToCall = carrierType.GetMethods().FirstOrDefault((x) => x.Name == "EarlyJobInit" && x.Parameters.Count == 0 && x.IsStatic && x.IsPublic);`

In `JobsILPostProcessor`. 

We must have a specially named function called `EarlyJobInit` that is static and public.

mbaker3 replied:
oh gross. ok. Can you leave a comment to that affect? Maybe pointing to the type or documentation that describes this hidden interface that we're conforming too?

jkeon replied:
Added comments and pointed to the type, no docs exist though.

jkeon committed:
Adding back in the indexer

jkeon committed:
Adding TODO reference

jkeon committed:
Better checks

jkeon commented:
Updated based on your comments. 

Quite a few of them I had responses for that require your input or an discussion via a call. 


> Can you check off what is complete in that issue? Or if it makes more sense move the other pending work to new issues.

Yes, updated and checked off everything except for the one remaining effort which is more involved. #86 now refers here and is about going through the code base to create issues for the specific post entities 1.0 upgrade work to do.

mbaker3 commented on Scripts/AssemblyInjections/Unity.Collections/FixedListInternalExtension.cs:
```diff
@@ -12,7 +12,6 @@ namespace Anvil.Unity.Collections
     /// access to function.
     /// </summary>
-    [BurstCompatible]
     public static unsafe class FixedListInternalExtension
```
Ok, I misunderstood how `[BurstCompile]` worked. That's fine.

I think comments might be helpful. It's fair enough that the IDE will warn you when calling non-burstable code from a bursted context but it doesn't really help when you're writing bursted code and poking around the codebase to see what you can find. Until you type it out (or read the whole class) you won't know if a piece of existing functionality that you've found will be compatible.

Maybe that doesn't happen often enough to justify writing comments and I'm inventing work here but I feel like, while writing bursted code, I'm often asking myself "can I use this method/type?"

jkeon replied:
The counter-argument to comments is that it might be too general and the underlying code changes and whoever does that forgets to update the comment at the top of the file? 

In the scenario you're describing, I feel like the effort to type it out (which is pretty fast with autocomplete) to get the Burst warning in the IDE is the same or less then opening things up and reading comments which would be more to maintain? 

If we really wanted to signify to the developer that these are Burst safe things to use, I would argue that we should note that in the Class name or Method name instead?

mbaker3 replied:
Let's try without the comments for now and see how it goes. 

When I'm devising a solution I'm not going to type out and check every method/type as I go. I'm going to collect a path or collection of types that "will work" and then go to implement.

Adding burst to all the type and method names is insane but if you're considering that I'll just drop this topic. You've backed me away from the ledge...

mbaker3 commented on Scripts/Runtime/Entities/Job/FloodSetComponentJob.cs:
```diff
@@ -29,7 +30,7 @@ public FloodSetComponentJob(ComponentTypeHandle<T> typeHandle, T value)
-        public unsafe void Execute(ArchetypeChunk chunk, int chunkIndex, int firstEntityIndex)
+        public unsafe void Execute(in ArchetypeChunk chunk, int unfilteredChunkIndex, bool useEnabledMask, in v128 chunkEnabledMask)
         {
             UnsafeCollectionUtil.FloodSetBuffer(chunk.GetComponentDataPtrRW(ref m_TypeHandle), 0, chunk.Count, m_Value);
```
Because the Query is the input here I disagree. If I wanted to apply to disabled I'd specify that in the query.

If this were a low level job where you're passing in the chunk yourself then I'd agree but since it's operating on queries we should obey its request.

jkeon committed:
Accounting for enabled

jkeon committed:
Including disabled entities and components

jkeon committed:
Adding comments for why a function exists

jkeon commented:
@mbaker3 Additional comments made and/or changes made. Back over to you!

jkeon committed:
Adding last comment
